Network Socket Statistics

deepikabhavnani opened the pull request:
### Description

This is the initial version of stats implementation in network layer.
SocketStats Class is added to collect and provide the statistics information. In this phase only socket information is collected and max sockets that can be recorded at any time are configurable through 'MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT'

Network statistics can be enabled through a macro MBED_NW_STATS_ENABLED

More information on design is captured in #8743

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

deepikabhavnani commented:
CC @SeppoTakalo @lauri-piikivi @sg- @SenRamakri

adbridge commented:
@deepikabhavnani, thank you for your changes. A review has been requested from the following:
@ARMmbed/mbed-os-core @ARMmbed/mbed-os-ipcore

Commit:
Resolved warning "call_in inherits implicit virtual"

lauri-piikivi commented on components/wifi/esp8266-driver/ESP8266Interface.cpp:
```diff
@@ -71,6 +76,11 @@ ESP8266Interface::ESP8266Interface(PinName tx, PinName rx, bool debug)
     _esp.setTimeout();
+
+#ifdef MBED_NW_STATS_ENABLED
+    _interface_stats.log_stack_addr((uint32_t)this);
```
stack_addr is a bit confusing name, change to stack_instance? Because stack_addr is easily confused with the IP address of the stack

lauri-piikivi commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
````diff
@@ -0,0 +1,186 @@
+|#Object:<Module_name>Stats stats;|     |#Object:EMACInterfaceStats stats;|   |#Object:LWIPStackStats stats;|
+`---------------------------------'     `---------------------------------'   `-----------------------------'
+```
+
````
Should it be IPStackStats instead of LWIP... We have multiple stacks

deepikabhavnani replied:
LWIPStackStats for LWIP stack, other stacks will need to have similar class and linked list entry

SeppoTakalo replied:
Why would the stats be implementation specific and not API specific?

`LWIPStack` is not Mbed OS API.
`NetworkStack` is

lauri-piikivi commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+
+## StatsList Class
+An abstract class used to create a linked list entry of every module reporting statistics data. Static API's in StatsList class can be used to read statistics data stored in linked list.
+
```
We leave it t the reader of the stats to understand if the values have changed -- it makes it hard for it to optimise the data communications to server? Could there be a flag or hash of data, that is easy to check if optimisations needs to be done?

deepikabhavnani replied:
Flag to indicate any parameter in stats structure has changed? We can add that but I will leave that to future extension of stats

lauri-piikivi commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
````diff
@@ -0,0 +1,186 @@
+}mbed_stats_type_t;
+```
+1. Statistic Info Structure - Structure used to get statistics data as binary blob from the StatsList class.
+
````
is the blob here with some encoding or just a "blob"?

deepikabhavnani replied:
Data is not encoded in any form

lauri-piikivi commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+
+EMAC interface class and LWIP stack together provided all information required for ethernet stattistics. `EMACInterfaceStats` and `LWIPStackStats` classes provide mechanism to log all statistics information.
+API's (`log_xxx`) provided by `EMACInterfaceStats` should be used by `EMACInterface` source to log interface related information such as IP address, netmask, etc.
+API's (`log_xxx`) provided by `LWIPStackStats` should be used by `LWIP` source to log network stack information such as tcp and udp sockets open, data packets transmitted, data packets received, etc.
```
I have a problem with the term log. To me it is a non-changing thing, if something is logged it is recorded at a timepoint (commonly) and will not be changed afterwards. We do not log something at timepoint, and we overwrite the value in the stats on every "logging". Should it be more neutral write_xxx or store_xxx

lauri-piikivi commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
````diff
@@ -0,0 +1,186 @@
+    stack_stats_t tcp;
+}nw_stack_stats_t;
+```
+
````
I think it might be better, for statistics and security information, that the send packets and received packets are reported per socket. Since the statistics is already a linked list, a socket when opened would register it's own status in the list, update the values, and the stats remain in that list when the socket is closed. Additional information would be the dest_addr, dest_port and time of open.

deepikabhavnani replied:
Do you mean adding entry in link list for each open socket? Do we have limit on number of sockets, we need to see the memory impact if all sockets are open.

SeppoTakalo replied:
All stacks limit the number of open sockets and that is configurable through JSON.
By default LwIP is limited to 4 sockets.

lauri-piikivi commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+    uint32_t pkt_recv;         /* Received packets. */
+    uint16_t sock_open;        /* Number of sockets opened */
+}stack_stats_t;
+
```
should it be clearly written that the data is cumulative during the lifetime of the module and its statistics

SeppoTakalo replied:
This structure is usefull.
But as the other statistics API, we should have one call to get this structure alone.

deepikabhavnani replied:
We have once call to get this structure. Please see the example in design https://github.com/ARMmbed/mbed-os/pull/8592/files#diff-a0a9946f63fe46b6b9aa87283d32b059R132. 

User is not aware of linked list, user facing API's are same as other existing statistics. User can allocate an array of struct and pass same count to fetch equivalent number of element in linked list.

lauri-piikivi commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
````diff
@@ -0,0 +1,186 @@
+    uint8_t connect_status;
+    bool dhcp;
+}nw_emac_info_t;
+```
````
Why do we put this data in stats, when the reader can read it from the Network API if and when it needs?
Seems unnecessary to make a duplicate o f this data in stats

--- if we really want to duplicate the information ---
stack_addr is instance, addr is a bit confusing term for network stack
A network interface can have both IPv4 and IPv6 at the same time. 
For IPv6 there is the default_router (gateway in ipv4) and prefix (something like netmask).

SeppoTakalo replied:
This structure is 166 bytes and contains data that can already be asked from NetworkInterface API.
Too much RAM usage for no benefit.

deepikabhavnani replied:
> Why do we put this data in stats, when the reader can read it from the Network API if and when it needs?

Reading from Network API is not essentially same as reading from statistics. We have thread information available and API's to fetch thread stack size as user API's but still same information is captured in statistics.

SeppoTakalo commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+
+Mbed OS supports various network interfaces and stacks, and selection of each can vary based on different targets, modules and default configurations. Users cannot easily fetch underlying network interface/stack information and that too without accessing the underlying creation object. To fetch network information without prior knowledge of interfaces supported and selected `StatsList` class will ...
+
+`StatsList` - An abstract class used to create a linked list entry of module reporting statistics data. `StatsList` class will also provide static API's to read statistics data stored in linked list.
```
I don't understand the need of this statslist at all.

Current Mbed OS device statistics API is very simple and clean. It has defined structures and one API call per structure to get those statistics.

This new API is not following the same approach.

This is creating linked list of statistics, which all are different type. Making no sense to link those together.

Also, as we are pushing the data collection to application side, we need to keep in mind that Mbed OS does not have network interface objects. Only applications can create network interfaces. So providing a list of interfaces for application that itself have instantiated makes not much sense.
"Here are the interfaces you have created"

deepikabhavnani replied:
> Mbed OS does not have network interface objects. Only applications can create network interfaces. So providing a list of interfaces for application that itself have instantiated makes not much sense.

We report stack statistics / thread statistics or any other statistics of objects created by application. Application allocating extra memory / creating additional threads does know about those elements does that mean even stack / thread statistics serve no purpose?

SeppoTakalo commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+    STATS_NW_INTERFACE_EMAC,
+    STATS_NW_INTERFACE_WIFI,
+    STATS_NW_INTERFACE_MESH,
+    STATS_NW_INTERFACE_CELLULAR,
```
This is the part that I don't agree.
You are forcing the data collection application to be aware of each of the interface types.
If they are aware, they can also query the information directly from those interfaces.

deepikabhavnani replied:
We can pass the interface type as start of interface structure, but this is essentially for filtering. When we have multiple stats reported by same linked list, user can request for specific type only.

See `get_each(stats_info_t *stats, int count)` vs `get_each(stats_info_t *stats, mbed_stats_type_t type, int count);`  https://github.com/ARMmbed/mbed-os/pull/8592/files#diff-a0a9946f63fe46b6b9aa87283d32b059R111

SeppoTakalo commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+    STATS_NW_INTERFACE_CELLULAR,
+
+    STATS_NW_STACK_LWIP,
+    STATS_NW_STACK_WIFI,
```
What does this mean?

LWIP is implementation of `NetworkStack`
Some WiFi drivers are also implementations of `NetworkStack` and implementations of `WiFiInterface`

But what is WIFI stack? We don't have such API

deepikabhavnani replied:
Stack stats and interface stats are separated, though we do not have full WIFI stack, but all socket information needed for stack stats will be added to separate structure then interface stats

SeppoTakalo commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+
+#### API's
+
+1. static int get_each(stats_info_t *stats, int count);
```
By fetch, do you mean that the data is going to be copied into the `stats_info_t *stats` and.. linked by how.. This is very confusing API. How much do we need to allocate memory to get the statistics and can it change in the runtime?

Could we just return the link to the head of the linked list that we already have?

deepikabhavnani replied:
It is implemented the same way as other stats, user need to allocate memory for array of `stats_info_t` structure based on the `count` requested

deepikabhavnani replied:
User is not aware of linked list, API's report stats as an array to structure. Memory allocated by user will be equivalent to `size of (stats_info_t ) * count` . Users do not need to allocate memory for RAM buffers of network interface and stack stats

SeppoTakalo commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+
+## Ethernet Statistics
+
+EMAC interface class and LWIP stack together provided all information required for ethernet stattistics. `EMACInterfaceStats` and `LWIPStackStats` classes provide mechanism to log all statistics information.
```
Why do you try to design here how drivers internally gathers the data.

We should not be dictating the API design into the drivers.

SeppoTakalo commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
```diff
@@ -0,0 +1,186 @@
+    uint32_t stack_addr;
+    stack_stats_t udp;
+    stack_stats_t tcp;
+}nw_stack_stats_t;
```
I don't see the purpose of having UDP and TCP stats separately. There might be other protocols as well (NB-IoT, LoRA). We should aim to be more generic.

It is already known in the design phase of the application which protocol they use.

deepikabhavnani replied:
Please suggest what all should be added/removed from this structure. We can have common count for udp/tcp sockets if that is more meaningful

SeppoTakalo commented on docs/design-documents/platform/Statistics/NetworkStatistics.md:
````diff
@@ -0,0 +1,186 @@
+}nw_stack_stats_t;
+```
+
+Various functions are added to log information related to each specific element like `log_ip`, `log_gateway`, `log_dhcp`... `log_stack_addr()` should be implemented by both interface and stack layer to know which interface uses which underlying stack in case of multiple networks. Apart from specific API's some generic API's are available for network stack layer like `lwip_log_add` and `lwip_log_s ...
````
This forces a new API into network interfaces and network stacks.

SeppoTakalo reviewed (changes requested):
I cannot approve this design.

This is way too far away from what current Mbed OS statistics APIs are.

Only usable thing in this PR is the `stack_stats_t` which contains packet counters.

If that is what have been requested, then we should have designed API similar to existing statistics APIS.

For example:
```
/** Network statistics */
typedef struct {
    uint32_t pkt_xmit;         /**< Transmitted packets. */
    uint32_t pkt_recv;         /**< Received packets. */
    uint16_t sock_open;        /**< Number of sockets opened */
} network_stats_t;
/** Get the network statistics */
void mbed_stats_network_get(network_stats_t *stats);
```

This would match the existing [mbed_stats.h](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_stats.h) API.

I cannot understand why network statics have been done so overly complex way.

If the example proposal I just show is fine, it should have been asked from IP networking team to implement. If more complex statistics have been asked, then I'm not aware of the requirements and we should have started by creating a design that can be justified by the requirements document.

deepikabhavnani commented:
>If the example proposal I just show is fine, it should have been asked from IP networking team to implement. If more complex statistics have been asked, then I'm not aware of the requirements and we should have started by creating a design that can be justified by the requirements document.

This is not the right forum to discuss this, but I would leave a comment stating design was shared nearly 3 months ago and numerous meetings request were made to discuss that as well

Commit:
This is the initial version of stats implementation in network layer.

SocketStats Class is added to collect and provide the statistics information.
In this phase only socket information is collected and max sockets that can
be recorded at any time are configurable through 'MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT'

Network statistics can be enabled through a macro MBED_NW_STATS_ENABLED

More information on design is captured in #8743

cmonr commented:
@deepikabhavnani This needs a rebase.

Commit:
Update udp test cases to verify network socket statistics

Cases Updated are:
1. UDPSOCKET_ECHOTEST_NONBLOCK
2. UDPSOCKET_OPEN_CLOSE_REPEAT
3. UDPSOCKET_OPEN_LIMIT

Commit:
Update tcp test cases to verify network socket statistics

Cases Updated are:
1. TCPSOCKET_ECHOTEST_NONBLOCK
2. TCPSOCKET_OPEN_CLOSE_REPEAT
3. TCPSOCKET_OPEN_LIMIT

deepikabhavnani commented:
@SeppoTakalo @lauri-piikivi @SenRamakri - Please review

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers - Travis astyle failures related to trailing spaces, there is no clear indication for user that error is because of extra space.
```
-     *                  If the number of sockets on the system is less than or equal to count, 
+     *                  If the number of sockets on the system is less than or equal to count,
```

Commit:
Code/bug after testing
1. Fixing astyle and docs
2. Extra mutex lock was removed
3. Bytes are updated when send/recv > 0 and not in case of -ve error
4. Review comments
5. Guard statistics implementation in test with MBED_NW_STATS_ENABLED

kegilbert commented on features/netsocket/SocketStats.h:
```diff
@@ -0,0 +1,115 @@
+
+    /** Create an socket statictics object
+     *
+     *  Applicatiopn users must not create class object.
```
```suggestion
     *  Application users must not create class objects.
```

kegilbert commented on features/netsocket/SocketStats.h:
```diff
@@ -0,0 +1,115 @@
+    /** Create an socket statictics object
+     *
+     *  Applicatiopn users must not create class object.
+     *  Class object will be created by entities reporting network statistics.
```
```suggestion
     *  The class object will be created by entities reporting network statistics.
```

kegilbert commented on features/netsocket/SocketStats.h:
```diff
@@ -0,0 +1,115 @@
+     *
+     *  Applicatiopn users must not create class object.
+     *  Class object will be created by entities reporting network statistics.
+     *  Application can fetch network statistics using static `mbed_stats_socket_get_each` API
```
```suggestion
     *  Applications can fetch network statistics using the static `mbed_stats_socket_get_each` function
```

kegilbert commented on features/netsocket/SocketStats.h:
```diff
@@ -0,0 +1,115 @@
+     *  Applicatiopn users must not create class object.
+     *  Class object will be created by entities reporting network statistics.
+     *  Application can fetch network statistics using static `mbed_stats_socket_get_each` API
+     *  without creatin an object.
```
```suggestion
     *  without creating an object.
```

kegilbert replied:
Sorry for the large blob of comments...Github doens't support multi-line edits yet, all very minor nits

deepikabhavnani replied:
@kegilbert - Github allows you to merge also one by one :-) Will squash them all

kegilbert commented on features/netsocket/SocketStats.h:
```diff
@@ -0,0 +1,115 @@
+#include "hal/ticker_api.h"
+
+#ifndef MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT
+#define MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT      10
```
Should this just live in the config system completely with a default value of 10?

deepikabhavnani replied:
Need to add this to the config system

kegilbert commented on features/netsocket/SocketStats.h:
```diff
@@ -0,0 +1,115 @@
+    static size_t mbed_stats_socket_get_each(mbed_stats_socket_t *stats, size_t count);
+
+    /** Add entry of newly created socket in statistics array.
+        @Note: Entry in the array will be maintained even after socket is closed.
```
Multi-line editing is being a pain ->

>Entry in the array will be maintained even after socket is closed. It will be over-written for closed sockets when socket entries in `MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT` exceed.

to (and @AnotherButler can probably phrase it better):

>Entries in the array will be maintained even after the socket is closed. Closed sockets will be overwritten first in the case of `MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT` being exceeded.

kegilbert commented on features/netsocket/SocketStats.h:
```diff
@@ -0,0 +1,115 @@
+    static uint32_t _size;
+
+    /** Internal function to scan the array and get position of element in the list.
+        This API locks the mutex and next API updating the entry in array
```
```suggestion
        This API locks the mutex and the next API call updating the entry in the array
```

kegilbert commented on features/netsocket/SocketStats.cpp:
```diff
@@ -0,0 +1,163 @@
+                }
+            }
+        }
+        if (-1 == position) {
```
Very minor, swapping logic order for consistency (variable -> LHS, constant -> RHS)

```suggestion
        if (position == -1) {
```

deepikabhavnani replied:
I prefer constants in left, helps in catching `=` instead of `==`

kegilbert commented on features/netsocket/SocketStats.cpp:
```diff
@@ -0,0 +1,163 @@
+        // Determine which entry in the list shall be over-written
+        for (uint32_t j = 0; j < MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT; j++) {
+            if (SOCK_CLOSED == _stats[j].state) {
+                if ((0 == oldest_time) || (oldest_time < _stats[j].last_change_tick)) {
```
Very minor, swapping logic order for consistency (variable -> LHS, constant -> RHS)

```suggestion
                if ((oldest_time == 0) || (_stats[j].last_change_tick > oldest_time)) {
```

deepikabhavnani replied:
Same thing, I prefer constants on left side of equation so they are caught when accidentally trying to over-write

kegilbert reviewed (commented):
@deepikabhavnani 
Pretty much just minor doc tweaks and very minor style changes on my end, fantastic work!

kegilbert commented on features/lwipstack/LWIPStack.h:
```diff
@@ -450,7 +450,7 @@ class LWIP : public OnboardNetworkStack, private mbed::NonCopyable<LWIP> {
      *  @return         0 on success, negative error code on failure
      */
-    nsapi_error_t call_in(int delay, mbed::Callback<void()> func);
+    virtual nsapi_error_t call_in(int delay, mbed::Callback<void()> func);
```
What was the reasoning for making this virtual? Just curious, didn't see this change referenced elsewhere in the PR.

deepikabhavnani replied:
Resolved warning https://github.com/ARMmbed/mbed-os/pull/8592/commits/022ebd14e28a6b8491e05e7efa4e94fd5b6b8490

kegilbert commented on features/netsocket/TCPSocket.cpp:
```diff
@@ -89,11 +100,13 @@ nsapi_error_t TCPSocket::connect(const SocketAddress &address)
 
     /* Non-blocking connect gives "EISCONN" once done - convert to OK for blocking mode if we became connected during this call */
     if (ret == NSAPI_ERROR_IS_CONNECTED && blocking_connect_in_progress) {
+    _socket_stats.stats_update_socket_state(this, SOCK_CONNECTED);
```
This should let astyle pass.

```suggestion
        _socket_stats.stats_update_socket_state(this, SOCK_CONNECTED);
```

deepikabhavnani replied:
Its already fixed.. travis CI was not getting triggered automatically, hence error is still present

kegilbert replied:
Alrighty, sorry for the redundant comment! Overheard @cmonr mentioning something about Travis issues today, may be related.

Commit:
Add config options for socket statistics

1. MBED_CONF_NSAPI_SOCKET_STATS_ENABLE to enable the statistics
2. MBED_CONF_NSAPI_SOCKET_STATS_MAX_COUNT max sockets statistics cached

deepikabhavnani commented:
@SeppoTakalo - Unit test are failing with linker error. Any idea what needs to be done in unit test for this ? 
```
22:27:07 [linux] /builds/ws/mbed-os-ci_unittests/unitTest-1229/mbed-os/features/netsocket/TCPServer.cpp:28: undefined reference to `SocketStats::stats_update_socket_state(Socket const*, socket_state)'
22:27:07 [linux] 
22:27:07 [linux] libfeatures-netsocket-TCPServer.MbedOS.a(TCPServer.cpp.o): In function `TCPServer::accept(TCPSocket*, SocketAddress*)':
22:27:07 [linux] 
22:27:07 [linux] /builds/ws/mbed-os-ci_unittests/unitTest-1229/mbed-os/features/netsocket/TCPServer.cpp:57: undefined reference to `SocketStats::stats_update_peer(Socket const*, SocketAddress const&)'
22:27:07 [linux] 
22:27:07 [linux] /builds/ws/mbed-os-ci_unittests/unitTest-1229/mbed-os/features/netsocket/TCPServer.cpp:58: undefined reference to `SocketStats::stats_update_socket_state(Socket const*, socket_state)'

```

Commit:
Add socketstats stub functions for unittest and addressed reviews

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers  - Astyle complaints about 4 tab spacing in cmake files? Should we restrict this check for C/H/CPP files?

deepikabhavnani commented on UNITTESTS/features/netsocket/InternetSocket/unittest.cmake:
```diff
@@ -25,4 +25,6 @@ set(unittest-test-sources
   stubs/stoip4_stub.c
   stubs/ip4tos_stub.c
+  stubs/SocketStats_Stub.cpp
+  stubs/SocketStats_Stub.cpp
```
```suggestion

```

0xc0170 commented:
> @ARMmbed/mbed-os-maintainers - Astyle complaints about 4 tab spacing in cmake files? Should we restrict this check for C/H/CPP files?

Fixed 3 days ago. It was restricted but grep command did not "anchor" .c so cmake was also selected. If you rebase on top of master (at least to get travis config file in), should fix cmake styling changes there. Sorry about that

deepikabhavnani commented:
@0xc0170 - Thanks. Rebased on top of master now.

SeppoTakalo commented on features/netsocket/SocketStats.h:
```diff
@@ -0,0 +1,114 @@
+    SOCK_LISTEN,                    /**< Socket is listening for incoming connections */
+} socket_state;
+
+typedef struct {
```
Add `/** ... */` Doxygen blocks for both of these structs as well. Members are already documented.

deepikabhavnani replied:
Done

SeppoTakalo commented on features/netsocket/TCPServer.cpp:
```diff
@@ -20,10 +20,12 @@ using mbed::Callback;
 
 TCPServer::~TCPServer()
 {
+    _socket_stats.stats_update_socket_state(this, SOCK_CLOSED);
```
Move this into the `InternetSocket::close()` because socket can be closed way before it is destructed.

SeppoTakalo commented on features/netsocket/TCPServer.cpp:
```diff
@@ -20,10 +20,12 @@ using mbed::Callback;
 
 TCPServer::TCPServer()
 {
+    _socket_stats.stats_new_socket_entry(this);
```
I think this can be moved into `InternetSocket` constructor.

SeppoTakalo commented on features/netsocket/TCPSocket.cpp:
```diff
@@ -28,17 +29,26 @@ TCPSocket::TCPSocket(TCPSocket *parent, nsapi_socket_t socket, SocketAddress add
 
 nsapi_protocol_t TCPSocket::get_proto()
 {
+    _socket_stats.stats_update_proto(this, NSAPI_TCP);
```
This is not in the correct place.

` _socket_stats.stats_update_proto(this, NSAPI_TCP);` should be in the contructor.
You cannot change the socket type after creating it.

SeppoTakalo commented on features/netsocket/TCPSocket.cpp:
```diff
@@ -28,17 +29,26 @@ TCPSocket::TCPSocket(TCPSocket *parent, nsapi_socket_t socket, SocketAddress add
+nsapi_error_t TCPSocket::close()
+{
+    _socket_stats.stats_update_socket_state(this, SOCK_CLOSED);
+    return InternetSocket::close();
```
`_socket_stats.stats_update_socket_state(this, SOCK_CLOSED);` can be called in the `InternetSocket::close()` as it is same for all socket types.

Therefore we don't need this `TCPSocket::close()` method.

deepikabhavnani replied:
~~I do not have socket instance (this) in InternetSocket, hence cannot move `stats_update_socket_state` calls to InternetSocket..~~ 

~~Same is true for open and entry functions~~

SeppoTakalo commented on features/netsocket/TCPSocket.h:
```diff
@@ -55,6 +55,15 @@ class TCPSocket : public InternetSocket {
+     *
+     *  @return         NSAPI_ERROR_OK on success, negative error code on failure
+     */
+    virtual nsapi_error_t close();
```
Remove this, and move the stat call into `InternetSocket`

SeppoTakalo commented on features/netsocket/UDPSocket.cpp:
```diff
@@ -20,20 +20,32 @@
+nsapi_error_t UDPSocket::close()
+{
+    _socket_stats.stats_update_socket_state(this, SOCK_CLOSED);
+    return InternetSocket::close();
```
Same is in the `TCPSocket` case. Move the call into `InternetSocket` so we don't need these protocol specific `::close()` methods.

SeppoTakalo commented on features/netsocket/UDPSocket.cpp:
```diff
@@ -20,20 +20,32 @@
 nsapi_protocol_t UDPSocket::get_proto()
 {
+    _socket_stats.stats_update_proto(this, NSAPI_UDP);
+    _socket_stats.stats_update_socket_state(this, SOCK_OPEN);
```
Same as in TCPSocket: type should be set in the constructor. State should be set in the `::open()` call.

SeppoTakalo reviewed (changes requested):
Small changes suggested but otherwise looks OK.

Commit:
Edit SocketStats.h

Edit file.

deepikabhavnani commented:
@SenRamakri  @SeppoTakalo @lauri-piikivi - Please review the updates

SeppoTakalo commented on features/netsocket/TCPServer.cpp:
```diff
@@ -52,7 +53,8 @@ nsapi_error_t TCPServer::accept(TCPSocket *connection, SocketAddress *address)
             _stack->socket_attach(socket, &Callback<void()>::thunk, &connection->_event);
-
+            _socket_stats.stats_update_peer(this, *address);
+            _socket_stats.stats_update_socket_state(this, SOCK_OPEN);
```
Should these two lines actually be
```
            _socket_stats.stats_update_peer(connection, *address);
            _socket_stats.stats_update_socket_state(connection, SOCK_OPEN);
```
So it is not `this` but the new `connection` that we want to update the stats for.

SeppoTakalo replied:
And after `accept()` the TCPSocket should be in `SOCK_CONNECTED` state. Not open.

SeppoTakalo reviewed (approved):
Looks OK.

One bugfix proposed. Either fix it now, or create follow up PR once this goes in.
I assume this is in a hurry, so probably should go into testing now.
@0xc0170 Can you CI?

0xc0170 commented:
Testing might fail or be aborted, we are investigating few failures we are seeing

/morph build

0xc0170 commented:
Only IAR fails with an internal error

```
Compile: /usr/local/iccarm-7.80.1/bin/iccarm --no_wrap_diagnostics -e --diag_suppress=Pa050,Pa084,Pa093,Pa082 -Oh --enable_restrict --cpu Cortex-M4F --thumb --dlib_config DLib_Config_Full.h --c++ --no_rtti --no_exceptions --guard_calls --no_static_destruction -DBOOTLOADER_ADDR=0x0 -DBOOTLOADER_SIZE=0x23000 -DAPPLICATION_ADDR=0x23000 -DAPPLICATION_SIZE=0x5d000 -DMBED_ROM_START=0x0 -DMBED_ROM_SIZE=0x80000 -DTARGET_MCU_NRF52832 -DCMSIS_VECTAB_VIRTUAL -D__MBED__=1 -DTARGET_LIKE_MBED -DDEVICE_PORTINOUT=1 -DTARGET_RTOS_M4_M7 -DMBED_TRAP_ERRORS_ENABLED=1 -DTARGET_NRF52832 -DCOMPONENT_PSA_SRV_EMUL=1 -DDEVICE_SERIAL_ASYNCH=1 -D__CMSIS_RTOS -D__CORTEX_M4 -DTARGET_SDT52832B -DBOARD_PCA10040 -DDEVICE_ITM=1 -DTARGET_NRF5x -DDEVICE_USTICKER=1 -DMBED_ALL_STATS_ENABLED -DDEVICE_I2C_ASYNCH=1 -DTARGET_CORTEX_M -DTARGET_LIKE_CORTEX_M4 -DTARGET_M4 -DFEATURE_BLE=1 -DCOMPONENT_PSA_SRV_IMPL=1 -DDEVICE_SPI_ASYNCH=1 -DDEVICE_LPTICKER=1 -DTOOLCHAIN_IAR -DMBED_HEAP_STATS_ENABLED=1 -DMBED_TICKLESS -DNRF52 -DTARGET_CORTEX -DDEVICE_I2C=1 -DDEVICE_PORTOUT=1 -DTARGET_SOFTDEVICE_S132_FULL -DTARGET_SOFTDEVICE_COMMON -DDEVICE_SERIAL=1 -D__FPU_PRESENT=1 -DDEVICE_PORTIN=1 -DTARGET_RELEASE -DTARGET_SDK_14_2 -DTARGET_NORDIC -DTARGET_NRF52 -DDEVICE_SYSTICK_CLK_OFF_DURING_SLEEP=1 -DDEVICE_TRNG=1 -DMBED_STACK_STATS_ENABLED=1 -D__MBED_CMSIS_RTOS_CM -DDEVICE_SLEEP=1 -DMBED_BUILD_TIMESTAMP=1542715360.44 -DDEVICE_SPI=1 -DCOMPONENT_NSPE=1 -DCMSIS_VECTAB_VIRTUAL_HEADER_FILE="cmsis_nvic.h" -DDEVICE_INTERRUPTIN=1 -DDEVICE_ANALOGIN=1 -DDEVICE_PWMOUT=1 -DDEVICE_FLASH=1 -DARM_MATH_CM4 -DDEVICE_SERIAL_FC=1 -f /builds/ws/mbed-os-build-matrix/3684/SDT52832B_IAR/sources/mbed-os/BUILD/tests/SDT52832B/IAR/.includes_0e89859847ddc6fc5f942feddc2be65c.txt --preinclude=/builds/ws/mbed-os-build-matrix/3684/SDT52832B_IAR/sources/mbed-os/BUILD/tests/SDT52832B/IAR/mbed_config.h --dependencies BUILD/tests/SDT52832B/IAR/features/netsocket/SocketStats.d -l BUILD/tests/SDT52832B/IAR/features/netsocket/SocketStats.s.txt -o BUILD/tests/SDT52832B/IAR/features/netsocket/SocketStats.o ./features/netsocket/SocketStats.cpp
[DEBUG] Return: 3
[DEBUG] Output: Internal error: [Front end]: assertion failed at: "7.80/core/translator/compiler_core/src/parser/edg/decl_inits.c", line 2031
```

Commit:
Resolve IAR build issue

IAR not able to assign default zero value in array and complaints of
internal error as below:

[ERROR] Internal error: [Front end]: assertion failed at:
"..\..\Translator\compiler_core\src\parser\edg\decl_inits.c", line 2031

Commit:
Corrected TCPServer reporting as per review comments

deepikabhavnani commented:
> Only IAR fails with an internal error

Added commit to fix IAR issue

0xc0170 commented:
/morph build

cmonr commented:
Please take a look at the build failures.

deepikabhavnani commented:
> Please take a look at the build failures.

Memory issue for all the low end devices :-( . Will have a look

Commit:
Make sure no memory overhead if statistics are disabled

deepikabhavnani commented:
Trying build to see if IAR issues are resolved.

/morph build

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers - Used CI build time to verify the IAR fix for low end devices and it passed for most of them. 
Build was aborted to avoid CI testing.

Please trigger the build and test as per CI availability. Thanks

0xc0170 commented:
/morph build

0xc0170 commented:
@deepikabhavnani Can you review unittests errors, one of errors I spot there

```
mbed-os-ci_unittests/unitTest-1249/mbed-os/features/netsocket/InternetSocket.cpp:77:5: error: ‘_socket_stats’ was not declared in this scope
```

deepikabhavnani commented:
Removed guard around object creation as its taken care in member functions. https://github.com/ARMmbed/mbed-os/commit/b196bd29efa5732ee33d2b42cf099a18ba934bb4#diff-ac8311f272aa21f3b23380f8f64b6c08R179

This should fix unit test

0xc0170 commented:
CI started

0xc0170 commented:
@deepikabhavnani Unit test still fails, please review

0xc0170 commented:
CI started

0xc0170 commented:
Errored, looks like SingletonPtr does not include header file that it needs (NULL definition) ? Comes from `SocketStats_Stub.cpp` file.

What I noticed, these stubs have lot of warnings, we should eliminate them

NirSonnenschein commented:
@deepikabhavnani : note this also needs a rebase before it is re-run.

bulislaw commented:
All the PRs need to be engineering ready (marked as "needs: CI") by the end of the day (Austin time). Otherwise it won't make 5.11 and will need to come in the next release (5.12 for features, 5.11.1 for fixes and new platforms).

Commit:
Resolve build issue for unittest

Error Resolved here is:
SingletonPtr.h:93:13: error: ‘NULL’ was not declared in this scope

cmonr commented:
PR updated. Restarting CI.

cmonr commented:
@deepikabhavnani Looks like there are still unittest issues.

Commit:
Add stats stub functions to rest of the cmake files

deepikabhavnani commented:
@cmonr - Fixed now.. Was able to test on linux machine

cmonr commented:
Thanks @deepikabhavnani. CI restarted.

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers - Please have a look

Stopped the test run stuck from 4 hours.

<1s
/builds/ws/mbed-os-ci_mbed2-build-GCC_ARM@2/749/XADOW_M0_GCC_ARM_749@tmp/durable-ae8d361d: No space left on device

deepikabhavnani commented:
/builds/ws/mbed-os-ci_mbed2-build-IAR/746/USENSE_IAR_746@tmp/durable-44ce7b4b: No space left on device

cmonr commented:
Restarted jobs with failed results.

Somehow, other jobs seem to not have been hit with the disk storage issue. Fingers crossed.
